### ISSUE TITLE
[no-meaningless-void-operator] Clashes with no-confusing-void-expression

### ISSUE THREAD

**Issue by jakubmazanec** [typescript-eslint/typescript-eslint#4507]
I'm not sure if this is a bug or I'm just misunderstanding the rules, but no-meaningless-void-operator and no-confusing-void-expression can contradict each other.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-confusing-void-expression": ["error", {"ignoreVoidOperator": true}],
    "@typescript-eslint/no-meaningless-void-operator": ["error"]
  }
}
```

```TS
type NextFn = () => void;

function middleware(next: NextFn): void {
    return void next(); // void operator shouldn't be used on void; it should convey that a return value is being ignored
}
```

But when I remove the void operator:

```TS
type NextFn = () => void;

function middleware(next: NextFn): void {
    return next(); // void expressions returned from a function must be marked explicitly with the `void` operator
}
```

Same with promises:

```TS
type NextFn = () => void;

async function middleware(next: NextFn): Promise<void> {
    return void next(); // void operator shouldn't be used on void; it should convey that a return value is being ignored
}
```

```TS
type NextFn = () => void;

async function middleware(next: NextFn): Promise<void> {
    return next(); // void expressions returned from a function must be marked explicitly with the `void` operator
}
```

**Expected Result**

I'm not sure what version should be correct, but I think this one:

```TS
type NextFn = () => void;

function middleware(next: NextFn): void {
    return void next(); // no error
}
```

But I don't want to turn no-meaningless-void-operator rule off - so should there be an exception for this situation, i.e. that using void operator on a function returning void while returning from a function that also returns void is okay?

**Comment by bradzacher** [comment]
cc @JoshuaKGoldberg as I know you care about both rules.

unfortunately we can't make the rules "communicate" because all lint rules are completely isolated - but maybe there's something else we can do here.

**Comment by JoshuaKGoldberg** [comment]
I care about all rules (that aren't formatting related). ❤️‍🔥

**Comment by JoshuaKGoldberg** [comment]
Err, is this not the solution?

```ts
function middleware1(next: NextFn): void {
  next();
}
```

If the root of the issue is that `return ... next()` is unnecessary because it's a `void`, why does it need to be returned in the first place?

**Comment by jakubmazanec** [comment]
Thanks for the answer! That is the obvious (well, not to me, apparently) and simple solution. I think it didn't occur to me because the actual code had stuff like early return, e.g.:

```TS
function middleware(next: NextFn): void {
  if (something) {
    return void next();
  }

  return void next();
}
```

This has one slight drawback: the function will be one line longer (or two, because I like to have a blank line before `return`):
```TS
function middleware(next: NextFn): void {
  if (something) {
    next();

    return;
  }

  next();
}
```

But if there isn't other solution (I can't think of anything), I think we can close this.

**Comment by JoshuaKGoldberg** [comment]
👍 closing away!
